packages/next-swc/crates/next-core/js/src/overlay/internal/container/Errors.tsx

```diff
@@ -130,221 +130,287 @@ const enum DisplayState {
-  const hasServerError = readyErrors.some((err) =>
-    ["server", "edge-server"].includes(getErrorSource(err.error) || "")
-  );
-
-  // TODO for now it's already closable, but in future we might want to block users from using a broken app
-  // const isClosable = !isLoading && !hasTurbopackErrors && !hasServerError;
-  const isClosable = true;
-
-  const defaultTab =
-    TAB_PRIORITY.find(
-      (tabId) =>
-        ({
-          [TabId.TurbopackErrors]: turbopackErrors.length > 0,
-          [TabId.TurbopackWarnings]: turbopackWarnings.length > 0,
-          [TabId.RuntimeErrors]: hasErrors,
-        }[tabId])
-    ) ?? TabId.RuntimeErrors;
-
-  const [selectedTab, setSelectedTab] = React.useState<string>(defaultTab);
+  const [readyErrors, _isLoading] = useResolvedErrors(errors)
 
-  React.useEffect(() => {
-    if (defaultTab === TabId.TurbopackErrors) {
-      setSelectedTab(TabId.TurbopackErrors);
-    }
-  }, [defaultTab]);
+  const tabs = TABS.map((tab) => ({
+    id: tab.id,
+    autoOpen: tab.autoOpen,
+    severity: tab.severity,
+    icon: tab.icon,
+    message: tab.message,
+    pluralTitle: tab.pluralTitle,
+    singularTitle: tab.singularTitle,
+    as: tab.as,
+    items: tab.items({ issues, readyErrors }),
+  })).filter((tab) => tab.items.length > 0)
 
-  const onlyHasWarnings = !hasErrors && !hasTurbopackErrors;
+  // Selected tab, null means it's closed
+  const [selectedTab, setSelectedTab] = React.useState<string | null>(null)
 
-  const [stateDisplayState, { fullscreen, minimize, hide }] = useDisplayState(
-    onlyHasWarnings ? DisplayState.Minimized : DisplayState.Fullscreen
-  );
-  let displayState = stateDisplayState;
+  // Toast is hidden
+  const [hidden, setHidden] = React.useState(false)
 
-  if (!isClosable) {
-    displayState = DisplayState.Fullscreen;
-  }
+  // Already seen issue ids, to auto open the dialog on new errors
+  const [seenIds, setSeenIds] = React.useState(() => new Set())
+
+  React.useEffect(() => {
+    const newSeenIds = new Set()
+    let change = false
+    let autoOpen = false
+    // When the selected tab disappears we will go to another important tab or close the overlay
+    if (selectedTab && !tabs.some((tab) => tab.id === selectedTab)) {
+      const otherImportantTab = tabs.find((tab) => tab.autoOpen)
+      if (otherImportantTab) {
+        setSelectedTab(otherImportantTab.id)
+      } else {
+        setSelectedTab(null)
+      }
+    } else {
+      autoOpen = true
+    }
+    // When there is a new item we open the overlay when autoOpen is set
+    for (const tab of tabs) {
+      for (const item of tab.items) {
+        newSeenIds.add(itemHash(item))
+        if (!seenIds.has(itemHash(item))) {
+          change = true
+          setHidden(false)
+          if (autoOpen && tab.autoOpen) {
+            setSelectedTab(tab.id)
+          }
+        }
+      }
+    }
+    if (change || newSeenIds.size !== seenIds.size) setSeenIds(newSeenIds)
+  }, [selectedTab, tabs, seenIds])
```

Comment:
This will execute every render because `tabs` is not in a memo, is that intentional?